src/core/net/ip6.cpp

```diff
@@ -1269,7 +1269,7 @@ Error Ip6::HandleDatagram(Message &aMessage, Netif *aNetif, const void *aLinkMes
         hopLimit = header.GetHopLimit();
         aMessage.Write(Header::kHopLimitFieldOffset, hopLimit);
 
-        if (aFromNcpHost && nextHeader == kProtoIcmp6)
+        if (nextHeader == kProtoIcmp6)
```

Comment:
I understand the goal is to ensure the ICMP6 filter is applied to the message that are passed from user/host to Thread (which I think is the original goal of this `if` block).

But the earlier changing of `aFromNcpHost` behavior and removing of the check here has a side-effect that now the same filter will also be applied to messages received by device (this code path is used for handling all messages whether received by device over Thread radio or given by user).

In terms of ICMP6:
- Within OT core we process echo request/response and `AddressResolver` processes `kTypeDstUnreach`) which are included in `sForwardICMPTypes` (so they are allowed and not filtered) 
- But we also have public OT ICMP6 API `otIcmp6RegisterHandler` which allows a user of OT to register a specific handler to process received ICMP6 message. So the filtering can now potentially apply the message types that can be used here as well.

I think the change in this PR in fine and make sense to filter in all cases. But just wanted to point this out so that we are aware of this change in behavior (in case it comes up later).

Reply:
`sForwardICMPTypes` seems to already contain all ICMPv6 message types that Thread network use. So, this change should be fine. A better way of filtering ICMPv6 packet may be dropping the messages via `ip6table` firewall rules.

Meanwhile, `aFromNcpHost` change seems to also have impact on the `OPENTHREAD_CONFIG_UNSECURE_TRAFFIC_MANAGED_BY_STACK_ENABLE` logic below. What are the possible side-effects? @gjc13 @bukepo

Reply:
This flag is there solely for a legacy Weave feature called Remote Passive Rendezvous. This features requires some 15.4 packets to be sent by a new device with out security enabled. The replied packets also need to disable security.

From the code it looks like the `aFromNcpHost` change will break this feature. Is this feature still important? In fact this is quite a weird behavior.

Reply:
IIUC, we don't need to check `aFromNcpHost` for `OPENTHREAD_CONFIG_UNSECURE_TRAFFIC_MANAGED_BY_STACK_ENABLE`. That is to say, all Thread devices can disable security if this feature is enabled. 

Thoughts?

Reply:
Security will be only disabled when the application explicitly opens an unsecure port and only for packets sent back. Though it may be still a security concern, it's required for the legacy Weave feature to work. Shall we ignore `aFromNcpHost` for `OPENTHREAD_CONFIG_UNSECURE_TRAFFIC_MANAGED_BY_STACK_ENABLE`?

Reply:
I agree with you guys, we need to address the situation under `OPENTHREAD_CONFIG_UNSECURE_TRAFFIC_MANAGED_BY_STACK_ENABLE` as well (in a new PR I guess).

I think as @simonlingoogle mentiond it should be fine to apply the port check indepdent of `aFromHost` flag. The only case I can think of is the for Thread MeshCoP (commissioning) but I think in this case the `AddUnsecurePort()` is called by OT core itself (e.g. from both `Joiner` and `JoinerRouter`) so filtering based on that should be fine.

----

> Should the `aFromHost` take place of the `aFromNcpHost`?

Yes. I fully agree. We should rename it (we can submit a rename PR later... I was considering `aFromUser` but `aFromHost` sounds better)

Reply:
So action items:

Get this PR merged to fix the ICMP6 issue.
A new PR to fix `OPENTHREAD_CONFIG_UNSECURE_TRAFFIC_MANAGED_BY_STACK_ENABLE`. This feature flag is quite tricky and I strongly suggest to sync with @bukepo offline to understand the background.
A new PR renaming `aFromNcpHost` and setting correct `fromHost` values.

Thoughts? @simonlingoogle @abtink

Reply:
Agree on action items. 

FYI @bukepo

Reply:
Agree with the action items.

For renaming, we should differentiate SentFromHost from ForwardedFromHost.

Reply:
I went back and tried to find the history of the change that added `UNSECURE_TRAFFIC_MANAGED_BY_STACK_ENABLE` behavior.
- Seems like it was added from https://github.com/openthread/openthread/pull/4782
- Here is the related code, which disables link security on an IPv6 message when the port matches even if security is explicitly requested for message the tx (by the caller): 
```c++
        if (aFromNcpHost && (nextHeader == kProtoTcp || nextHeader == kProtoUdp))
        {
            uint16_t sourcePort, destPort;

            SuccessOrExit(error = aMessage.Read(aMessage.GetOffset(), sourcePort));
            SuccessOrExit(error = aMessage.Read(aMessage.GetOffset() + sizeof(sourcePort), destPort));
            sourcePort = HostSwap16(sourcePort);
            destPort   = HostSwap16(destPort);

            //...

#if OPENTHREAD_CONFIG_UNSECURE_TRAFFIC_MANAGED_BY_STACK_ENABLE
            // check whether source port is an unsecure port
            if (Get<Filter>().IsUnsecurePort(sourcePort))
            {
                aMessage.SetLinkSecurityEnabled(false);
                otLogInfoIp6("Disabled link security for packet to %s", header.GetDestination().ToString().AsCString());
            }
#else
            OT_UNUSED_VARIABLE(sourcePort);
#endif
        }
```
- Thinking more on this, I don't think we want to relax and apply this rule to all IPv6 message.
- **I actually favor removing this whole feature `UNSECURE_TRAFFIC_MANAGED_BY_STACK_ENABLE`** (I had some concerns about this earlier -> https://github.com/openthread/openthread/pull/4782#discussion_r404992091)
- I don't recall why we needed this behavior. Probably for weave pairing. I think if there is such a use-case it would be better to manage it at the next layer and not add this in OT core(i.e., it is the job of the next layer to decide whether it wants to enable link security for an IPv6 message it is passing to Thread stack). If I recall correctly, that was how weave pairing worked with `wpantund`.